frontend/src/game.js

```diff
@@ -56,37 +248,227 @@ export class Game {
+                    this.cur_move.to
+                ]
+            })
+        );
+    }
+
+    // private
+    send_attack() {
+        let coords = this.find_unit(2);
+        let params = this.grid.hexes[coords.y][coords.x].unit.params;
+        this.socket.send (
+            JSON.stringify ({
+                "cmd": "change",
+                "type": "attack",
+                "coords": {
+                    "from": this.cur_move.to,
+                    "to": this.cur_attack
+                },
+                "changes": {
+                    "hurt": [
+                        {"x":coords.x,"y":coords.y,"unit":{
+                            "player":params.player,"hp":params.hp - 1,"attack":params.attack,"speed":params.speed
+                        }}
+                    ]
+                }
+            })
+        );
+    }
+
+
+    // Change field functions
+    // private
+    redraw_field(data) {
+        console.log('redraw field');
+        // bad switch, need to switch by data.type
+        switch(data.type) {
+            case 'move':
+                this.move_unit(data.coords[0].y, data.coords[0].x, data.coords[1].y, data.coords[1].x);
+            break;
+
+            case 'attack':
+                console.log('attack!!! charge!!!');
+                // animate attack
+                if (data.changes) {
+                    if (data.changes.hurt) {
+                        data.changes.hurt.forEach(el => {
+                            this.change_hex(el);
+                        });
+                    }
+                }
+            break;
+        }
+    }
+
+    // private
+    change_hex(data) {
+        console.log(data);
+        let hex = this.grid.hexes[data.y][data.x];
+        if (data.content) {
+            hex.set_content(data.content);
+        }
+        if (data.unit) {
+            hex.change_unit(data.unit);
+        }
+    }
+
+
+    // Actions with units
+    // private
     create_unit(texture, img_size, params, y, x) {
         let unit = new Unit(texture, img_size, params);
         this.grid.hexes[y][x].set_unit(unit);
         return unit;
     }
 
+    // private
     move_unit(from_y, from_x, to_y, to_x) {
         let from_hex = this.grid.hexes[from_y][from_x];
         let to_hex = this.grid.hexes[to_y][to_x];
         let unit = from_hex.unit;
+        if (from_hex.unit === null) {
+            throw new Error(`No unit in the cell to move from: ${JSON.stringify(from_hex.coords)}`);
+        }
+        if (to_hex.unit !== null) {
+            throw new Error(`Cell to move to is already occupied: ${JSON.stringify(to_hex.coords)}`);
+        }
 
         from_hex.unset_unit();
         to_hex.set_unit(unit);
     }
+
+    // dev
+    // private
+    find_unit(player) {
+        for (let y = 0; y < this.grid.row_n; y++) {
+            for (let x = 0; x < this.grid.col_n; x++) {
+                let unit = this.grid.hexes[y][x].unit;
+                if (unit) {
```

Comment:
In JS operators `&&` and `||` are short-circuited, so we can just write `if (unit && unit.params.player === player)`, and if the unit is `Null`, the second part of `&&` would not be evaluated:
```js
> b = {c : "qwe"}
{ c: 'qwe' }
> if (b && b.b) { console.log("b has field with name b"); }
undefined
> if (b && b.c) { console.log("b has field with name c"); }
b has field with name c
undefined
```

Reply:
Ok, will try

Reply:
done!